src/Mono.Android/Microsoft.Android.Runtime/ManagedValueManager.cs

```diff
@@ -20,52 +22,80 @@ class ManagedValueManager : JniRuntime.JniValueManager
 {
 	const DynamicallyAccessedMemberTypes Constructors = DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors;
 
-	Dictionary<int, List<IJavaPeerable>>?   RegisteredInstances = new Dictionary<int, List<IJavaPeerable>>();
+	readonly Dictionary<int, List<ReferenceTrackingHandle>> RegisteredInstances = new ();
+	readonly ConcurrentQueue<IntPtr> CollectedContexts = new ();
+
+	bool disposed;
+
+	static readonly SemaphoreSlim bridgeProcessingSemaphore = new(1, 1);
 
 	static Lazy<ManagedValueManager> s_instance = new (() => new ManagedValueManager ());
 	public static ManagedValueManager GetOrCreateInstance () => s_instance.Value;
 
-	ManagedValueManager ()
+	unsafe ManagedValueManager ()
 	{
+		// There can only be one instance of ManagedValueManager because we can call JavaMarshal.Initialize only once.
+		var mark_cross_references_ftn = RuntimeNativeMethods.clr_initialize_gc_bridge (&BridgeProcessingStarted, &BridgeProcessingFinished);
+		JavaMarshal.Initialize (mark_cross_references_ftn);
 	}
 
-	public override void WaitForGCBridgeProcessing ()
+	protected override void Dispose (bool disposing)
 	{
+		disposed = true;
+		base.Dispose (disposing);
 	}
 
-	public override void CollectPeers ()
+	void ThrowIfDisposed ()
 	{
-		if (RegisteredInstances == null)
+		if (disposed)
 			throw new ObjectDisposedException (nameof (ManagedValueManager));
+	}
 
-		var peers = new List<IJavaPeerable> ();
+	public override void WaitForGCBridgeProcessing()
+	{
+		bridgeProcessingSemaphore.Wait();
+		bridgeProcessingSemaphore.Release();
+	}
```

Comment:
Nit, missing spaces before parentheses

```suggestion
	public override void WaitForGCBridgeProcessing ()
	{
		bridgeProcessingSemaphore.Wait ();
		bridgeProcessingSemaphore.Release ();
	}
```